### ISSUE TITLE
azcopy --recursive copy from source Linux server to File share doesn't work

### ISSUE THREAD

**Issue by rbavery** [Azure/azure-storage-azcopy#629]
##### Note: The version is visible when running AzCopy without any argument
10.2.1
### Which platform are you using? (ex: Windows, Mac, Linux)
Linux
### What command did you run?

```
./azcopy copy --recursive ../ARD-all-nebraska-2002-2019/ <SAS string copied from Azure File Storage>
```
### What problem was encountered?
```
Log file is located at: /home/rave/.azcopy/6917c0c9-b2ea-e84b-51ec-ca9ae3f13f52.log

0 Done, 880 Failed, 424 Pending, 0 Skipped, 1304 Total, 


Job 6917c0c9-b2ea-e84b-51ec-ca9ae3f13f52 summary
Elapsed Time (Minutes): 0.0667
Total Number Of Transfers: 1304
Number of Transfers Completed: 0
Number of Transfers Failed: 1304
Number of Transfers Skipped: 0
TotalBytesTransferred: 0
Final Job Status: Failed

```

The log file is long and its hard to parse where the relevant error message is, but here is a section of it. I don't think the error is correct because the Resource should be found since I copied the SAS string from file explorer.
```
2019/10/01 19:55:47 ERR: [P#0-T#1289] error deleting the (incomplete) file https://cropmaskperm.file.core.windows.net/azureml-filestore-896933ab-f4fd-42b2-a154-0abb35dfb0b0/unpacked_ard_landsat_downloads/ARD-all-nebraska-2002-2019/LT05_CU_015009_20020822_20190111_C01_V01_SR.tar?<SAS removed>. Failed with error -> github.com/Azure/azure-storage-file-go/azfile.newStorageError, /home/vsts/go/pkg/mod/github.com/!azure/azure-storage-file-go@v0.5.0/azfile/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=ResourceNotFound) =====
Description=The specified resource does not exist.
RequestId:1008d3f3-101a-00a3-5d92-789dc8000000
Time:2019-10-01T19:55:47.1641420Z, Details: 
   Code: ResourceNotFound
   DELETE https://cropmaskperm.file.core.windows.net/azureml-filestore-896933ab-f4fd-42b2-a154-0abb35dfb0b0/unpacked_ard_landsat_downloads/ARD-all-nebraska-2002-2019/LT05_CU_015009_20020822_20190111_C01_V01_SR.tar?<SAS removed>
   User-Agent: [AzCopy/10.2.1 Azure-Storage/0.5.0 (go1.12; linux)]
   X-Ms-Client-Request-Id: [d8a6a496-7d54-4229-683b-a36ac9cd5dbe]
   X-Ms-Version: [2018-03-28]
   --------------------------------------------------------------------------------
   RESPONSE Status: 404 The specified resource does not exist.
   Content-Length: [223]
   Content-Type: [application/xml]
   Date: [Tue, 01 Oct 2019 19:55:46 GMT]
   Server: [Windows-Azure-File/1.0 Microsoft-HTTPAPI/2.0]
   X-Ms-Error-Code: [ResourceNotFound]
   X-Ms-Request-Id: [1008d3f3-101a-00a3-5d92-789dc8000000]
   X-Ms-Version: [2018-03-28]

```

### How can we reproduce the problem in the simplest way?
Not sure because it is so simple, so something might be different about either my source or destination. But you can try

Downloading azcopy, then

```
./azcopy login --tenant-id <tenant id from azure portal>
# get SAS url from file explorer
./azcopy copy --recursive ../ARD-all-nebraska-2002-2019/ <SAS string copied from Azure File Storage> # fails
```

### Have you found a mitigation/solution?
No

**Comment by adreed-msft** [comment]
Hi there, @rbavery. ResourceNotFound is a common way that Azure Files dismisses an authentication failure in a secure way. Please ensure that your SAS token has adequate permissions to access the container. I can't see any parameters of your SAS token, so I'm unaware of what permissions the token may have, etc.

**Comment by rbavery** [comment]
I checked all the permissions boxes when creating the SAS token, including READ, WRITE, DELETE, etc. Could it be something else? I will tray again.

**Comment by adreed-msft** [comment]
OK. Do you see any UPLOADFAILED markers within your log? That will contain the cause of the failure, typically.

**Comment by rbavery** [comment]
Yes for everything I trie dto upload there is this line 

`120670:2019/10/01 20:01:44 ERR: [P#0-T#1298] UPLOADFAILED: ../ARD-all-nebraska-2002-2019/LT05_CU_012009_20020818_20190111_C01_V01_ST.tar : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: a68c08ea-801a-006a-1393-782025000000`

But this path does exist.

```
(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ ls
azcopy  ThirdPartyNotice.txt
(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ pwd
/home/rave/forge_scratch/azcopy_linux_amd64_10.2.1
(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ ls ..
ARD                         dowload_urls.csv      unpacked_ARD
ARD-all-nebraska-2002-2019  espa-bulk-downloader
azcopy_linux_amd64_10.2.1   fakedir
(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ 
```

**Comment by adreed-msft** [comment]
According to the azure files documentation

The following file names are not allowed: LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, LPT9, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, PRN, AUX, NUL, CON, CLOCK$, dot character (.), **and two dot characters (..).**

I would suspect this applies to folders as well. In the upcoming version of AzCopy, the use of relative paths in this fashion should be resolved. Please try uploading from an immediate path to the directory, rather than using the shortcut ..

**Comment by rbavery** [comment]
That's all true but isn't the source of the error I think. I'm still getting errors when using absolute paths, and when I don't use symbolic links

```
(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ ./azcopy copy --recursive /scratch/rave/ARD-all-nebraska-2002-2019/ https://cropmaskperm.file.core.windows.net/azureml-filestore-896933ab-f4fd-42b2-a154-0abb35dfb0b0/unpacked_ard_landsat_downloads/?<SAS removed>
```

**Comment by rbavery** [comment]
from the new log file, the error is the same and says it doesn't exist. But it does exist, since I can list it

```
121237:2019/10/01 20:49:54 ERR: [P#0-T#1183] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_015007_20020806_20190517_C01_V01.jpg : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: bc890d36-b01a-0061-4b99-78db4e000000
(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ ls /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_015007_20020806_20190517_C01_V01.jpg
/scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_015007_20020806_20190517_C01_V01.jpg

```

**Comment by rbavery** [comment]
btw thanks for the quick response on these issues, I really appreciate it!

**Comment by adreed-msft** [comment]
Strange that it's still registering as not found. These errors wouldn't be local-- they're registering a service code; as a result, they're a service side kind of "not found".

It's possible that the share might not exist at all. Could you refresh storage explorer and take a look?

**Comment by adreed-msft** [comment]
Furthermore, there may be distinct permissions for creating folders on Azure File Shares.

**Comment by adreed-msft** [comment]
A third thought is that you may need to provide your SAS token list permissions as well.
Here's some of the code where we create parent directories:
```go
	if _, err := dirURL.GetProperties(ctx); err != nil {
		if err.(azfile.StorageError) != nil && (err.(azfile.StorageError)).Response() != nil &&
			(err.(azfile.StorageError).Response().StatusCode == http.StatusNotFound) { // At least need read and write permisson for destination
```

**Comment by adreed-msft** [comment]
There's a permission called create on SAS tokens. Does this help out?

**Comment by rbavery** [comment]
It definitely exists, I've refreshed storage explorer and confirmed this.

Not sure what your last comments get at, should I create a new folder rather than try to copy to an existing folder? I'll try that and check that folder's permissions when I create it.

**Comment by rbavery** [comment]
I don't see the permission "Create on SAS tokens" in Storage Explorer, only Create, Delete, Read, Write, and List.

I tried something slightly different after looking at the azcopy definition more closely, still doesn't work but now an error prints to the console

```
Authentication failed, it is either not correct, or expired, or does not have the correct permission -> github.com/Azure/azure-storage-file-go/azfile.newStorageError, /home/vsts/go/pkg/mod/github.com/!azure/azure-storage-file-go@v0.5.0/azfile/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=AuthenticationFailed) =====
Description=Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
RequestId:e05f9870-601a-0062-08a7-783a2a000000
Time:2019-10-01T22:29:28.1840991Z, Details: 
   AuthenticationErrorDetail: The specified signed resource is not allowed for the this resource level
   Code: AuthenticationFailed
   GET https://cropmaskperm.file.core.windows.net/azureml-filestore-896933ab-f4fd-42b2-a154-0abb35dfb0b0/ard-partial/ARD-all-nebraska-2002-2019?<SAS removed>
   User-Agent: [AzCopy/10.2.1 Azure-Storage/0.5.0 (go1.12; linux)]
   X-Ms-Client-Request-Id: [3fd8a7a0-140d-4726-78dc-6431f42142ec]
   X-Ms-Version: [2018-03-28]
   --------------------------------------------------------------------------------
   RESPONSE Status: 403 Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
   Content-Length: [447]
   Content-Type: [application/xml]
   Date: [Tue, 01 Oct 2019 22:29:27 GMT]
   Server: [Microsoft-HTTPAPI/2.0]
   X-Ms-Error-Code: [AuthenticationFailed]
   X-Ms-Request-Id: [e05f9870-601a-0062-08a7-783a2a000000]



(geo) [rave@forge azcopy_linux_amd64_10.2.1]$ /scratch/rave/azcopy_linux_amd64_10.2.1/azcopy copy --recursive=True "/scratch/rave/ARD-all-nebraska-2002-2019/" "https://cropmaskperm.file.core.windows.net/azureml-filestore-896933ab-f4fd-42b2-a154-0abb35dfb0b0/ard-partial<SAS removed>"

```

I don't know why my authentication isn't working, I've logged in recently with azcopy with my only tenant id and I just generated that SAS string.

**Comment by JohnRusk** [comment]
I think Adele meant, did your SAS token, which you were using, include Create permissions (and it wouldn't hurt to test with a token that has read permission, to make sure it can read when it checks for parent directory existence - it's supposed to cope if the check fails, but no harm in testing with the token that has Read).

**Comment by rbavery** [comment]
Yes, I always check all the boxes when creating the SAS token. Not sure how to test with a token that has read permissions differently than how I am currently doing.

**Comment by rbavery** [comment]
I've also tried creating a new file share and a new blob store to copy the same directory and I've tried this on my Ubuntu laptop and a fedora server, neither are working and I'm still getting the authentication issue. This is the traceback for trying to copy to a blob, which it doesn't look like you can generate a SAS for.

```
→ azcopy copy --recursive=True "/home/rave/ARD-partial/" https://cropmaskperm.blob.core.windows.net/azcopy/ARD-partial/                                  
INFO: Scanning...
INFO: Using OAuth token for authentication.

Job 226f3376-c04f-7544-6f8e-5ef584c61801 has started
Log file is located at: /home/rave/.azcopy/226f3376-c04f-7544-6f8e-5ef584c61801.log

0 Done, 0 Failed, 1304 Pending, 0 Skipped, 1304 Total, 2-sec Throughput (Mb/s): 14.7794
Authentication failed, it is either not correct, or expired, or does not have the correct permission -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, /home/vsts/go/pkg/mod/github.com/!azure/azure-storage-blob-go@v0.7.0/azblob/zc_storage_error.go:42
===== RESPONSE ERROR (ServiceCode=AuthorizationPermissionMismatch) =====
Description=This request is not authorized to perform this operation using this permission.
RequestId:d52547ac-101e-009c-06ac-78556b000000
Time:2019-10-01T23:03:00.4896551Z, Details: 
   Code: AuthorizationPermissionMismatch
   PUT https://cropmaskperm.blob.core.windows.net/azcopy/ARD-partial/ARD-partial/LT05_CU_014008_20020921_20190517_C01_V01_BT.tar?timeout=901
   Authorization: REDACTED
   Content-Length: [808960]
   User-Agent: [AzCopy/10.2.1 Azure-Storage/0.7 (go1.12; linux)]
   X-Ms-Blob-Cache-Control: []
   X-Ms-Blob-Content-Disposition: []
   X-Ms-Blob-Content-Encoding: []
   X-Ms-Blob-Content-Language: []
   X-Ms-Blob-Content-Md5: []
   X-Ms-Blob-Content-Type: [application/x-tar]
   X-Ms-Blob-Type: [BlockBlob]
   X-Ms-Client-Request-Id: [a76ebcbb-9b6f-4007-7c2a-cbf75d81e8d7]
   X-Ms-Version: [2018-03-28]
   --------------------------------------------------------------------------------
   RESPONSE Status: 403 This request is not authorized to perform this operation using this permission.
   Content-Length: [279]
   Content-Type: [application/xml]
   Date: [Tue, 01 Oct 2019 23:03:00 GMT]
   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]
   X-Ms-Error-Code: [AuthorizationPermissionMismatch]
   X-Ms-Request-Id: [d52547ac-101e-009c-06ac-78556b000000]
   X-Ms-Version: [2018-03-28]

```

**Comment by JohnRusk** [comment]
Thanks for the detailed info Ryan. We (probably Adele) will look into it some more.  Probably won't be today though, I'm sorry.

**Comment by JohnRusk** [comment]
adoladeg-microsoft ignore bogus assignment of issue to you, I just clicked your name instead of Adeles!

**Comment by rbavery** [comment]
@adreed-msft any updates?

**Comment by rbavery** [comment]
I saw the restest label, I've retested and I still get the same error

```
(base) [rave@forge rave]$ /scratch/rave/azcopy_linux_amd64_10.2.1/azcopy copy --recursive /scratch/rave/ARD-all-nebraska-2002-2019/ https://cropmaskperm.file.core.windows.net/azureml-filestore-896933ab-f4fd-42b2-a154-0abb35dfb0b0/ard-partial?<SAS removed>
[1] 16939
[2] 16940
[3] 16941
[4] 16942
[5] 16943
[2]   Done                    se=2019-10-02T22%3A11%3A16Z
[3]   Done                    sp=rcwdl
[4]-  Done                    sv=2018-03-28
[5]+  Done                    sr=f
(base) [rave@forge rave]$ INFO: Scanning...

Job d9139a2e-15f8-5d46-5ae0-65791565cc5a has started
Log file is located at: /home/rave/.azcopy/d9139a2e-15f8-5d46-5ae0-65791565cc5a.log

0 Done, 1305 Failed, 0 Pending, 0 Skipped, 1305 Total,  


Job d9139a2e-15f8-5d46-5ae0-65791565cc5a summary
Elapsed Time (Minutes): 0.1
Total Number Of Transfers: 1305
Number of Transfers Completed: 0
Number of Transfers Failed: 1305
Number of Transfers Skipped: 0
TotalBytesTransferred: 0
Final Job Status: Failed

```

log file shows UPLOADFAILED errors
```
2019/10/03 18:45:24 ERR: [P#0-T#1302] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_016008_20020831_20190111_C01_V01.jpg : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: 6aad9af0-c01a-007b-331a-7aba91000000
2019/10/03 18:45:24 ERR: [P#0-T#1300] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_013007_20020930_20190111_C01_V01_ST.tar : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: bd12b2b1-f01a-002d-571a-7a4b7e000000
2019/10/03 18:45:24 ERR: [P#0-T#1301] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_012009_20020804_20190111_C01_V01_ST.tar : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: 05fae7e6-601a-0062-691a-7a3a2a000000
2019/10/03 18:45:24 ERR: [P#0-T#1297] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_013009_20020905_20190517_C01_V01_ST.tar : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: 5a86a2e5-101a-000a-781a-7a5cba000000
2019/10/03 18:45:24 ERR: [P#0-T#1298] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_012009_20020804_20190111_C01_V01_BT.tar : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: 68046a01-401a-00dc-2f1a-7a5253000000
2019/10/03 18:45:24 ERR: [P#0-T#1303] UPLOADFAILED: /scratch/rave/ARD-all-nebraska-2002-2019/LT05_CU_017009_20020826_20190517_C01_V01_ST.tar : 404 : 404 The specified resource does not exist.. When Creating parent directory. X-Ms-Request-Id: 0d8e0cca-d01a-0015-731a-7aefbe000000

```

**Comment by JohnRusk** [comment]
Thanks for the detailed info.  It might come in handy.

BTW, I had poorly worded that label. It actually was supposed to mean "Needs retest by the AzCopy developer team, to reproduce the issue".  I have reworded it to make that clearer. Nevertheless, in this instance, the upload errors you posted may be informative.

**Comment by rbavery** [comment]
Any update on this?

**Comment by JohnRusk** [comment]
Sorry, @rbavery , we've been caught up in getting release 10.3 out the door.  @adreed-msft Is this issue something you'd have time to look at in the next 24 hours?

**Comment by adreed-msft** [comment]
>[5]+  Done                    sr=f
BINGO!

You've got a signed resource for a file. This should be sr=s for share-level permissions. You can't create folders with a SAS for a specific file.

**Comment by JohnRusk** [comment]
Thanks Adele.  I'll close this now.  @rbavery, please reply if you need it re-opened.

**Comment by rbavery** [comment]
@adreed-msft I'm not sure what this means sorry, but came back to let you know that 10.3.1 seems to work for me when syncing a different folder to a new fileshare that doesn't have any folders in it. Was the problem that I was creating an empty folder and then a SAS for that empty folder (which was treated like a file?)
